### PR TITLE
feat: minimize unnecessary API requests to the API server for the configmap/secret get API

### LINKED ISSUE #120126: kubelet get configmap from kube-apiserver and etcd everytime
### What happened?

qps of configmap get request is too high in kube-apiserver.

when there are many configmaps in cluster mounted by running pod, the qps of configmap get request will be high and the request will also need to get configmap resource from etcd instead of returned from kube-apiserver cache.

### What did you expect to happen?

DO NOT get configmap from etcd. Get it from cache of kube-apiserver.

### How can we reproduce it (as minimally and precisely as possible)?

1. set kubelet configMapAndSecretChangeDetectionStrategy to Cache;
2. set kube-apiserver --v=3
3. create lots of configmaps mount to pods
4. check kube-apiserver metrics `sum(rate(apiserver_request_total{resource="configmaps",verb="GET"}[5m])) by (verb,resource,subresource)` 
5. check kube-apiserver log

### Anything else we need to know?

I have tested it in v1.22.13 and v1.28.0. The logic to set resourceversion=0 seems never be triggered

### Kubernetes version

<details>

```console
1.22.13, 1.28.0 
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by likakuli**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/120126

#### Special notes for your reviewer:
Only invoke AddReference first time when to sync same pod to minimize unnecessary API requests to the API server for the configmap/secret get API.

#### Does this PR introduce a user-facing change?

```release-note
--sync-frequency will not affect the update interval of volumes that use ConfigMaps or Secrets when the configMapAndSecretChangeDetectionStrategy is set to Cache. The update interval is only affected by node.alpha.kubernetes.io/ttl node annotation."
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/util/manager/cache_based_manager.go [hunk 1]
```diff
@@ -224,21 +224,15 @@ func (c *cacheBasedManager) RegisterPod(pod *v1.Pod) {
...
-			// are only in prev need to be completely removed. This unconditional
-			// call takes care of both cases.
-			c.objectStore.DeleteReference(prev.Namespace, name, prev.UID)
+	// To minimize unnecessary API requests to the API server for the configmap/secret get API
+	// only invoke AddReference the first time RegisterPod is called for a pod.
+	if prev == nil {
+		for name := range names {
+			c.objectStore.AddReference(pod.Namespace, name, pod.UID)
 		}
 	}
```
Comment (liggitt) [inline pkg/kubelet/util/manager/cache_based_manager.go:237]: this needs an else case

```go
} else {
  prevNames := c.getReferencedObjects(prev)
  // add new references
  for name := range names {
    !if prevNames.Has(name) {
      c.objectStore.AddReference(pod.Namespace, name, pod.UID)
    }
  }
  // remove dropped references
  for prevName := range prevNames {
    !if names.Has(prevName) {
      c.objectStore.DeleteReference(pod.Namespace, prevName, pod.UID)
    }
  }
}
```

File: pkg/kubelet/util/manager/cache_based_manager_test.go [hunk 2]
```diff
@@ -414,6 +414,7 @@ func TestCacheInvalidation(t *testing.T) {
 			{envVarNames: []string{"s3"}},
 		},
 	}
+	manager.UnregisterPod(podWithSecrets("ns1", "name1", s1))
```
Comment (liggitt) [inline pkg/kubelet/util/manager/cache_based_manager_test.go:417]: revert this line

File: pkg/kubelet/util/manager/cache_based_manager_test.go [hunk 3]
```diff
@@ -495,6 +496,7 @@ func TestCacheRefcounts(t *testing.T) {
 			{envVarNames: []string{"s5"}, envFromNames: []string{"s50"}},
 		},
 	}
+	manager.UnregisterPod(podWithSecrets("ns1", "name2", s1))
```
Comment (liggitt) [inline pkg/kubelet/util/manager/cache_based_manager_test.go:499]: revert this line

File: pkg/kubelet/util/manager/cache_based_manager_test.go [hunk 4]
```diff
@@ -520,6 +522,7 @@ func TestCacheRefcounts(t *testing.T) {
 
 	// Also check the Add + Update + Remove scenario.
 	manager.RegisterPod(podWithSecrets("ns1", "other-name", s1))
+	manager.UnregisterPod(podWithSecrets("ns1", "other-name", s1))
```
Comment (liggitt) [inline pkg/kubelet/util/manager/cache_based_manager_test.go:525]: revert this line

File: pkg/kubelet/util/manager/cache_based_manager.go (hunk 1 above)
Comment (likakuli) [reply pkg/kubelet/util/manager/cache_based_manager.go:237]: Is it possible for pod in same ns with same uid and name that volumes changed only with a pod update event?

File: pkg/kubelet/util/manager/cache_based_manager_test.go (hunk 2 above)
Comment (likakuli) [reply pkg/kubelet/util/manager/cache_based_manager_test.go:417]: i think there is a pod delete event in the actual world when to change pod volumes, for volumes in pod spec can't be modified directly.

Comment (liggitt) [reply pkg/kubelet/util/manager/cache_based_manager_test.go:417]: there are references other than volumes that can change after creation (like envFrom references)

File: pkg/kubelet/util/manager/cache_based_manager.go (hunk 1 above)
Comment (liggitt) [reply pkg/kubelet/util/manager/cache_based_manager.go:237]: volumes can't change, but other secret/configmap references can which this manager would have to handle. envFrom references in ephemeral containers are a good example

Comment (likakuli) [reply pkg/kubelet/util/manager/cache_based_manager.go:237]: understand

Comment (likakuli) [reply pkg/kubelet/util/manager/cache_based_manager.go:251]: kindly ping @liggitt. Can you please help to check whether it's ok now. thx

**Pull request merged**